### ISSUE TITLE
Plugin Loading Bypasses Package Dependency Validation

### ISSUE THREAD

**Issue by jondavidi** [cloud-custodian/cloud-custodian#4926]
A bug was introduced in #4195 which, for certain plugins, bypasses running dependency validation defined in `setup.py` using the `install_requires` flag for a given plugin. The affected plugins are the "core plugins" (`c7n_azure`, `c7n_gcp`, `c7n_kube`) which were changed to be loaded via hardcoded logic in `/cloud-custodian/c7n/resources/__init__.py`

Before the changes in #4195, every plugin was loaded via the `load_plugins()` method of the AWS resources registry object. This used the `pkg_resources` lookup functionality which inherently runs package validation as part of its loading logic. When the "core plugins" were changed to have hardcoded loading by directly importing the initialization function, this removed `pkg_resources` loading logic from the workflow, and thereby also removed any package dependency validation which would previously catch when package dependencies needed updating, such as in issue #2695.

The only way that dependency validation is being run now, is if a user includes the `C7N_EXTPLUGINS` environment variable for loading custom plugins, which triggers the original method of loading all plugins via `pkg_resources` module.

The impact is that for all users who have not been using the `C7N_EXTPLUGINS` environment variable since executing Custodian with the changes from #4195, there has not been any dependency validation being ran or enforced for the core plugins on their installation, and therefore their packages have the potential to run improperly or break.

**Comment by kapilt** [comment]
C7N_EXTPLUGINS is not a supported feature, its best effort, as it represents anti community and project enablement.

What was introduced was not a bug, but an optimization as pkg_resources is a hot path for serverless initialization.  Clearly this issue is not relevant to the vast majority of users that are using custodian in production without issue. The fact that it passes thousands of unit tests in ci is clear validation as well.

I've also talked to the python core maintainers and the direction I get is that pkg_resources isn't something we want to use. To that end, python 3.8 has introduced new std library modules (which pkg_resources isn't) to enable looking at installed package metadata.

verification of installation isn't a runtime consideration imo, its an install time consideration.

**Comment by jondavidi** [comment]
@kapilt What then would you suggest be used as an install-time mechanism for verifying dependencies? Right now nothing is enforcing it, and there's several dependency errors that get thrown in a typical `make install`. Before the changes in #4195, these dependency issues were addressed by the community and resolved because pkg_resources was enforcing validation, such as in #2695.

I definitely understand that we want the optimization for serverless initialization, but it seems like we're compromising on potential instability creeping it's way in if there isn't a mechnanism somewhere that's actually keeping dependencies in check.

The python 3.8 new libraries sounds like a great idea. I know that there's already features in 3.7 that do this (which may be what you're referring to) https://docs.python.org/3.7/library/importlib.html#module-importlib.resources
Is there a plan for moving over to use a new package metadata discovery service?

I believe there's a strong use-case for what `C7N-EXTPLUGINS` provides. Organizations in general have internal systems that can gain benefit from direct integration with Custodian. Doing this is not anti-community, because contributing a plugin for an internal system that others don't use would be useless to the project and would actually be something that does not legitimately have a use-case to be added to the core repo.

I think the reason that more organizations don't utilize this feature is simply because they don't know it exists....there's zero documentation on how to utilize it (which I would be happy to add).

And if this is not a supported feature, then I'm unsure as to why it was added. The reality is that since the functionality is there, it is inherently a feature that custodian supports (how well it is supported is another matter), and it is currently broken. I think it's worth discussing to try and find a solution to stabilize this feature because of the value it adds.

**Comment by kapilt** [comment]
external systems are intended to integrate with value_from on filters to decouple from the internal systems which often have vastly different slas, and to use invoke lambda/sfn for action integration. the extplugins is not intended for that and is anti pattern for those use cases. again extplugins may be removed in the future, it is not a supported feature.

**Comment by kapilt** [comment]
wrt to install time verification, using an installer that verifies the dependency chain is correct. the underlying issue here is the azure sdk is publishing unusable packages, which should be fixed upstream. what issues are you referencing in make install?

**Comment by jondavidi** [comment]
the issues I'm referencing during `make install`, is when packages are being installed there are several dependency errors raised in the stdout (but that don't block execution). These are the ones that my PR was fixing (azure and kubernetes version conflicts). It's only when actually loading these packages with `pkg_resources` at runtime that it throws the same errors, but in an execution blocking manner. 

Since the changes from #4195, these don't actually halt execution for users, as it previously would have, since its bypassing `pkg_resources` at runtime

**Comment by jondavidi** [comment]
To clarify, there's multiple dependency errors in the output for Azure and Kube that are not related to the azure sdk upstream issue

And therefore those fixable dependency requirements still cause blocking issues on their own with the use of `C7N_EXTPLUGINS`

**Comment by jondavidi** [comment]
@kapilt Understood about opening new ticket specifically for dependencies.

And regarding your `make install` output you linked to, lines 355-357 contain the dependency errors I've been referring to in this ticket, which break when they are loaded using `pkg_resources` methods.

For your reference from your output link (lines 355-357):
```
azure-cosmosdb-table 1.0.5 has requirement azure-storage-common~=1.1, but you'll have azure-storage-common 2.1.0 which is incompatible.

azure-cli-core 2.0.74 has requirement azure-mgmt-resource~=4.0, but you'll have azure-mgmt-resource 2.1.0 which is incompatible.

c7n-kube 0.1.1 has requirement kubernetes==9.0.0, but you'll have kubernetes 10.0.1 which is incompatible.
```

**Comment by kapilt** [comment]
the kube one is fixed in the release pr, the azure ones need upstream fixes from upstream before we can resolve fully.
